src/Strategy/Maven/DepTree.hs

```diff
@@ -0,0 +1,174 @@
+import Data.String (IsString)
+import Data.String.Conversion (toString, toText)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Void (Void)
+import DepTypes (
+  DepEnvironment (..),
+  DepType (MavenType),
+  Dependency (..),
+  VerConstraint (CEq),
+ )
+import Discovery.Walk (
+  WalkStep (WalkContinue),
+  findFileNamed,
+  walk,
+ )
+import Effect.Exec (AllowErr (Never), Command (..), Exec, execThrow)
+import Effect.Grapher (direct, edge, evalGrapher)
+import Effect.ReadFS (ReadFS, readContentsParser)
+import Graphing (Graphing, gmap)
+import Path (Abs, Dir, File, Path)
+import Path.IO (removeFile)
+import System.Random (randomIO)
+import Text.Megaparsec (
+  Parsec,
+  between,
+  many,
+  takeWhile1P,
+  try,
+ )
+import Text.Megaparsec.Char (space1)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+import Types (GraphBreadth (Complete))
+
+newtype DepTreeLabel
+  = BuildTag DepEnvironment
+  deriving (Eq, Ord, Show)
+
+outputFileName :: IsString a => a
+outputFileName = "fossa-deptree.dot"
+
+analyze :: (Has Exec sig m, Has ReadFS sig m, Has Diagnostics sig m, Has (Lift IO) sig m) => Path Abs Dir -> m (Graphing Dependency, GraphBreadth)
+analyze dir = do
+  -- We generate files with a random prefix so we don't pick up files from previous runs.
+  -- If we did pick them up, they're likely not enough of a conflict to worry about, but
+  -- it's better than not doing it.
+  randIdent <- abs <$> sendIO randomIO
+  _ <- context "Running maven 'dependency:tree' plugin" $ execThrow dir $ deptreeCmd randIdent
+  graphFiles <- context "Locating maven output files" $ findDepTreeOutputs dir randIdent
+  graphs <- context "Parsing output files" $ traverse (readContentsParser parseDotGraph) graphFiles
+  -- TODO: Will this work correctly in the face of exceptions?
+  sendIO $ traverse_ removeFile graphFiles
+  pure (buildGraph graphs, Complete)
+
+findDepTreeOutputs :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> Int -> m [Path Abs File]
+findDepTreeOutputs dir ident = execState @[Path Abs File] [] $
+  flip walk dir $ \_ _ files -> do
+    case findFileNamed (show ident <> outputFileName) files of
+      Nothing -> pure WalkContinue
+      Just file -> do
+        modify (file :)
+        pure WalkContinue
+
+buildGraph :: [DotGraph] -> Graphing Dependency
+buildGraph = gmap toDependency . foldr ((<>) . toGraph) mempty
+
+toDependency :: PackageId -> Dependency
+toDependency PackageId{artifactName, artifactVersion, buildTag} =
+  Dependency
+    { dependencyType = MavenType
+    , dependencyName = artifactName
+    , dependencyVersion = Just $ CEq artifactVersion
+    , dependencyLocations = []
+    , dependencyEnvironments = maybe [EnvProduction] ((: []) . toBuildTag) buildTag
+    , dependencyTags = mempty
+    }
+
+toGraph :: DotGraph -> Graphing PackageId
+toGraph DotGraph{rootNode, edgeList} = run . evalGrapher $ do
+  direct rootNode
+  for_ edgeList $ uncurry edge
```

Comment:
[no action] ugh this makes me want a better interface for building `Graphing` s -- similar to `AdjacencyMap`, where you could use, e.g., `edges` here, and `edge` has a better type

Reply:
I'm not sure how much we could really improve this by. There's always the trivial:

```hs
edges :: Ord t => [(t, t)] -> Graphing t -> Graphing t
```

Is that what you're talking about? Or do you mean a more drastic interface change?

Reply:
Or maybe a graph-shaped helper like this?

```hs
makeGraphing :: [t] -> [(t, t)] -> Graphing t
makeGraphing directs edges = ...
```

Reply:
I think `edge` and `direct` need to be simplified:
```hs
edge :: Ord a => a -> a -> Graphing a

direct :: a -> Graphing a

-- maybe with extra helpers like AdjacencyMap has, like you mentioned
directs :: Ord a => [a] -> Graphing a

edges :: Ord a => [(a,a)] -> Graphing a
```

..which would allow us to simplify `toGraph` here:
```hs
toGraph DotGraph{rootNode, edgeList} = direct rootNode <> edges edgeList
```

---

I think `Graphing.direct`/`Graphing.edge`/etc have near-zero use because they're not ergonomic at all, and I'd like to approximately mirror the design of `Algebra.Graph.AdjacencyMap` 's operations

Reply:
> I'd like to approximately mirror the design of Algebra.Graph.AdjacencyMap 's operations

Yeah, that's a good idea. Semantically, we're basically an "AM with directs" anyway.

Reply:
Yep exactly. `Graphing` is isomorphic to `AdjacencyMap (Node a)` where:

```hs
data Node a = Root | Node a
  deriving (Eq, Ord, Show)

type Graphing = AdjacencyMap (Node a)

direct :: Ord a => a -> Graphing a
direct = AM.edge Root (Node a)

edge :: Ord a => a -> a -> Graphing a
edge a b = AM.edge (Node a) (Node b)

..etc
```

but when writing it I preferred the clarity of it having its own distinct type and explicit set of direct deps